### PR TITLE
[mdal] [mesh] dataset groups

### PR DESCRIPTION
**Pull request by PeterPetrik**
Update to MDAL 0.0.5 and introducting dataset groups. A dataset group represents a quantity (e.g. depth or wind speed), a dataset represents values for one time for the quantity.

### CODE REVIEW

File: src/app/mesh/qgsmeshdatasetgrouptreeview.cpp [hunk 1]
```diff
@@ -231,55 +228,43 @@ void QgsMeshDatasetGroupTreeView::onSelectionChanged( const QItemSelection &sele
 
   if ( selected.isEmpty() )
   {
-    mActiveGroup = QString();
+    mActiveGroup = 0;
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshdatasetgrouptreeview.cpp]: Shouldn't we use something like -1 for no active group?

And emit activeGroupChanged() signal as well?

File: src/app/mesh/qgsmeshdatasetgrouptreeview.h [hunk 2]
```diff
@@ -130,8 +130,8 @@ class APP_EXPORT QgsMeshDatasetGroupTreeView : public QTreeView
 
     QgsMeshDatasetGroupTreeModel mModel;
     QgsMeshLayer *mMeshLayer = nullptr; // not owned
-    QMap<QString, QVector<int>> mGroups; // group name -> dataset indices
-    QString mActiveGroup;
+    QStringList mGroups; // group names
+    int mActiveGroup; // index of active group
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshdatasetgrouptreeview.h]: Initialize here

File: src/app/mesh/qgsmeshrendereractivedatasetwidget.cpp [hunk 3]
```diff
@@ -114,16 +124,16 @@ void QgsMeshRendererActiveDatasetWidget::updateMetadata( int datasetIndex )
...
-  int datasetIndex = -1;
-  if ( value < datasets.size() )
-  {
-    datasetIndex = datasets[value];
-  }
-  return datasetIndex;
+  int groupIndex = mDatasetGroupTreeView->activeGroup();
+  int datasetCount = mMeshLayer->dataProvider()->datasetCount( groupIndex );
+
+  if ( datasetCount > value && mMeshLayer && mMeshLayer->dataProvider() )
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrendereractivedatasetwidget.cpp]: few lines above the mMeshLayer->dataProvider() was used without any checks...

File: src/app/mesh/qgsmeshrendereractivedatasetwidget.h [hunk 4]
```diff
@@ -47,36 +48,36 @@ class APP_EXPORT QgsMeshRendererActiveDatasetWidget : public QWidget, private Ui
...
+    void updateMetadata( QgsMeshDatasetIndex datasetIndex );
 
   private:
-    int datasetIndex() const;
+    QgsMeshDatasetIndex datasetIndex() const;
 
     QgsMeshLayer *mMeshLayer = nullptr; // not owned
-    int mActiveScalarDataset = -1;
-    int mActiveVectorDataset = -1;
+    QgsMeshDatasetIndex mActiveScalarDataset ;
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrendereractivedatasetwidget.h]: extra space before the semicolon :-)

File: src/app/mesh/qgsmeshrendererscalarsettingswidget.h [hunk 5]
```diff
@@ -57,7 +58,7 @@ class APP_EXPORT QgsMeshRendererScalarSettingsWidget : public QWidget, private U
 
   public slots:
     //! Set active scalar dataset to be used
-    void setActiveDataset( int activeDatase );
+    void setActiveDataset( QgsMeshDatasetIndex activeDatase );
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrendererscalarsettingswidget.h]: s/activeDatase/activeDataset/

File: src/providers/mdal/qgsmdalprovider.cpp [hunk 6]
```diff
@@ -120,56 +118,83 @@ bool QgsMdalProvider::addDataset( const QString &uri )
...
 
-QgsMeshDatasetValue QgsMdalProvider::datasetValue( int datasetIndex, int valueIndex ) const
+QgsMeshDatasetMetadata QgsMdalProvider::datasetMetadata( QgsMeshDatasetIndex index ) const
 {
-  if ( datasetIndex >= mDatasets.length() )
+  DatasetGroupH group = MDAL_M_datasetGroup( mMeshH, index.group() );
+  if ( !group )
+    return QgsMeshDatasetMetadata();
+
+  DatasetH dataset = MDAL_G_dataset( group, index.dataset() );
```
Comment (wonder-sk) [inline src/providers/mdal/qgsmdalprovider.cpp]: Also return earlier if the returned dataset is invalid?

**Review by wonder-sk — COMMENTED** [review]
Looking good.

There were few MSVC compilation fixes from Jurgen that would be reverted with this update - see https://github.com/qgis/QGIS/commits/master/external/mdal

File: src/providers/mdal/qgsmdalprovider.cpp (hunk 6 above)
Comment (PeterPetrik) [reply src/providers/mdal/qgsmdalprovider.cpp]: done

**Review by wonder-sk — APPROVED** [review]

**Comment by nirvn** [comment]
@PeterPetrik , good job. This PR is now able to render a .grb2 file which failed to render until now. However, QGIS now crashes if I click on the contour panel's [ Recalculatr ] button.

Here's the dataset:
[gfs_3_20180708_0000_000.zip](https://github.com/qgis/QGIS/files/2209421/gfs_3_20180708_0000_000.zip)

Steps to reproduce:
- Launch QGIS
- Add the above-uploaded .grb2 dataset
- Open the style panel, switch to the contour tab
- Activate [x] Show contours
- Click on [ Recalculate ]
- *boom* QGIS dies

**Comment by saberraz** [comment]
@nirvn I think it was messed up due to this ticket:
https://github.com/lutraconsulting/MDAL/issues/19

**Comment by saberraz** [comment]
I too can confirm the issue.

**Comment by wonder-sk** [comment]
@nirvn This should be fixed in b3fb647ede

**Pull request merged**